### PR TITLE
fix Issue 20906 - unnecessary divide-by-zero errors when constant fol…

### PR DESCRIPTION
**Pull request by WalterBright**
…ding short circuits

### CODE REVIEW

**Comment by WalterBright** [comment]
```
CyberShadow/DAutoTest — Build failed
```
No clue in the log file why it failed.
```
ci/circleci: build — Your tests failed on CircleCI
```
Appears to not give enough time for the tests to complete.

**Comment by thewilsonator** [comment]
> No clue in the log file why it failed.

Look closer.

`Failed to download package taggedalgebraic from https://code.dlang.org/packages/taggedalgebraic/0.10.11.zip
`
Network error.

**Comment by WalterBright** [comment]
There are 6 `Failed to` messages in the log file, none of which match what you posted.

**Comment by MoonlightSentinel** [comment]
> Appears to not give enough time for the tests to complete.

No, your patch seems to trigger an infinite loop during self-compilation when building the `lexer` target.

**Comment by WalterBright** [comment]
Consider:

https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=4077946&isPull=true

It times out. I have no idea which test timed out.

**Comment by WalterBright** [comment]
How do I fix the tests to run serially instead of in parallel?

**Comment by MoonlightSentinel** [comment]
Pass `-j1` to `run.d`

**Comment by WalterBright** [comment]
Where's the command to run.d?

**Comment by WalterBright** [comment]
test/run.d does not appear to have a -J switch.

**Comment by MoonlightSentinel** [comment]
You can execute `rdmd run.d` inside of the `test` directory. Passing `--help` will also print some usage information.

**Comment by MoonlightSentinel** [comment]
The timeout apparently happens when running the unittests (`rdmd run.d -u`)

**Comment by WalterBright** [comment]
Thank you. What I find extremely frustrating about all this is NOBODY would accept these sorts of error messages from the compiler, yet the log files generated are ABSOLUTELY TERRIBLE in identifying WHAT IS BEING RUN and WHAT/WHERE THE ERROR IS.

> running the unittests

Which unit tests? Which component? dmd? druntime? phobos? something else? And why does the log file show a long list of test files being run if the error is somewhere else?

**Comment by WalterBright** [comment]
I just built dmd with unittests on and it does not hang.

**Comment by MoonlightSentinel** [comment]
> Which unit tests?

All tests inside of `test/unit`

The problem here is that all those tests are compiled into a single executable using the generated dmd. That compilation hangs and such the test runner never get's a chance to report errors.

But we should add some timeout detection to `run.d` to make such failures more obious.

**Comment by WalterBright** [comment]
blocking https://github.com/dlang/dmd/pull/11236

**Comment by MoonlightSentinel** [comment]
Running `dustmite` right now to find a reduced test case (but that might take some time)

**Comment by WalterBright** [comment]
Looking at test/unit:

1. No description of what it is, what it is for, etc.
2. No description of what the files are
3. No description of how to compile them (attempting to compile them just produces lots of compile errors)
4. files like `frontend.d` do not have a SINGLE COMMENT in them
5. no `main` function anywhere

https://issues.dlang.org/show_bug.cgi?id=20911

**Comment by WalterBright** [comment]
blocking https://github.com/dlang/dmd/pull/11236

**Comment by MoonlightSentinel** [comment]
`test/README.md` provides a short explanation on how to run the unittests. 
But it should also explain how to write such tests.

**Comment by WalterBright** [comment]
Ok, I see the readme. But it has a major failing - there are two compilers involved here. One is the compiler being used to build run.d, the other is the compiler being tested. The readme gives no clue.

How `run.d` fits in with `Makefile` there is no clue. `Makefile`, naturally, has no documentation saying what it's for.

**Comment by wilzbach** [comment]
> But it has a major failing - there are two compilers involved here. One is the compiler being used to build run.d, the other is the compiler being tested. The readme gives no clue.

You're supposed to use your bootstrap compiler like for `build.d` or more correctly `rdmd` (it's the first line of the script). rdmd will use your system's bootstrap compiler, so I really don't understand all the confusion. It's just `./run.d` in your console - like any other script and the testsuite runner will take care of everything else like the Makefile did and find your newly compiled DMD binary for you and use that, compile other parts of the testsuite runner etc.

Btw it's been there for about two years and IIRC about a year since the Makefile forwards to run.d. The fact that you haven't noticed that it's the way the testsuite is executed shows how smooth the transition was.

> How run.d fits in with Makefile there is no clue. Makefile, naturally, has no documentation saying what it's for.

It's legacy and only used by old CIs and packagers and basically just forwards to run.d.
There was a bit of a transition period to the newer run.d, but I guess by now all CIs could be changed to use run.d directly and at least that Makefile could be removed.

**Comment by WalterBright** [comment]
@wilzbach thank you. Please insert the text of your comments into run.d / Makefile

Not being able to find the code that causes the problem is still a huge problem. The design of the test/runnable etc. is the way it is specifically to make it easy to find the offending piece of code. With the test/unit making one giant program, it makes things a LOT harder.

I cringe whenever a problem appears that is not in test/runnable /compilable /fail-compilation because I know I'm in for many, many hours looking for it. I'm already over a full day into this one and am still nowhere.

**Comment by WalterBright** [comment]
@jacob-carlborg  since you wrote these tests, can you please have a look and perhaps find what piece of code in the test is resulting in failure of the compiler? Thanks!

**Comment by Geod24** [comment]
@WalterBright : You don't get any output anywhere because DMD enters an infinite busy loop, so all the CIs end up timing out. And if the unittests aren't enough, just take a look at buildkite, there are some projects passing, some failing.

One of the failing projects is `dub`, so I checked out this PR, and used it to build `dub`. Since `dub` can be build via a D script or `dub` itself, it takes an extra step to test the compiler. You have to use the `-v` option to get what command line is used.
```console
% dub build -v --compiler=../install/osx/bin/dmd
[...]
Performing "debug" build using ../install/osx/bin/dmd for x86_64.
Target '/Users/geod24/projects/dlang/dub/.dub/build/application-debug-posix.osx.darwin-x86_64-dmd_2092-F034260D70B7CBC3C9D827805FB05AB9/dub' doesn't exist, need rebuild.
dub 1.21.0+commit.52.g376ff585: building configuration "application"...
Using pkg-config to resolve library flags for libcurl.
../install/osx/bin/dmd -c -of.dub/build/application-debug-posix.osx.darwin-x86_64-dmd_2092-F034260D70B7CBC3C9D827805FB05AB9/dub.o -debug -g -w -version=DubUseCurl -version=DubApplication -version=Have_dub -Isource/ source/app.d source/dub/commandline.d source/dub/compilers/buildsettings.d source/dub/compilers/compiler.d source/dub/compilers/dmd.d source/dub/compilers/gdc.d source/dub/compilers/ldc.d source/dub/compilers/utils.d source/dub/dependency.d source/dub/dependencyresolver.d source/dub/description.d source/dub/dub.d source/dub/generators/build.d source/dub/generators/cmake.d source/dub/generators/generator.d source/dub/generators/sublimetext.d source/dub/generators/targetdescription.d source/dub/generators/visuald.d source/dub/init.d source/dub/internal/git.d source/dub/internal/libInputVisitor.d source/dub/internal/sdlang/ast.d source/dub/internal/sdlang/exception.d source/dub/internal/sdlang/lexer.d source/dub/internal/sdlang/package.d source/dub/internal/sdlang/parser.d source/dub/internal/sdlang/symbol.d source/dub/internal/sdlang/token.d source/dub/internal/sdlang/util.d source/dub/internal/utils.d source/dub/internal/vibecompat/core/file.d source/dub/internal/vibecompat/core/log.d source/dub/internal/vibecompat/data/json.d source/dub/internal/vibecompat/data/serialization.d source/dub/internal/vibecompat/data/utils.d source/dub/internal/vibecompat/inet/path.d source/dub/internal/vibecompat/inet/url.d source/dub/package_.d source/dub/packagemanager.d source/dub/packagesupplier.d source/dub/packagesuppliers/fallback.d source/dub/packagesuppliers/filesystem.d source/dub/packagesuppliers/maven.d source/dub/packagesuppliers/package.d source/dub/packagesuppliers/packagesupplier.d source/dub/packagesuppliers/registry.d source/dub/platform.d source/dub/project.d source/dub/recipe/io.d source/dub/recipe/json.d source/dub/recipe/packagerecipe.d source/dub/recipe/sdl.d source/dub/semver.d source/dub/version_.d -vcolumns
Serializing composite type BuildRequirements which has no serializable fields
Serializing composite type BuildOptions which has no serializable fields
source/dub/packagesuppliers/maven.d(82,10): Deprecation: module std.xml is deprecated - Will be removed from Phobos in 2.101.0. If you still need it, go to https://github.com/DigitalMars/undeaD
^C # Hangs forever
```

Now all one have to do is to copy paste that command, run it in a debugger, wait a bit, and CTRL+C

```console
% lldb -- ../install/osx/bin/dmd -c -of.dub/build/application-debug-posix.osx.darwin-x86_64-dmd_2092-F034260D70B7CBC3C9D827805FB05AB9/dub.o -debug -g -w -version=DubUseCurl -version=DubApplication -version=Have_dub -Isource/ source/app.d source/dub/commandline.d source/dub/compilers/buildsettings.d source/dub/compilers/compiler.d source/dub/compilers/dmd.d source/dub/compilers/gdc.d source/dub/compilers/ldc.d source/dub/compilers/utils.d source/dub/dependency.d source/dub/dependencyresolver.d source/dub/description.d source/dub/dub.d source/dub/generators/build.d source/dub/generators/cmake.d source/dub/generators/generator.d source/dub/generators/sublimetext.d source/dub/generators/targetdescription.d source/dub/generators/visuald.d source/dub/init.d source/dub/internal/git.d source/dub/internal/libInputVisitor.d source/dub/internal/sdlang/ast.d source/dub/internal/sdlang/exception.d source/dub/internal/sdlang/lexer.d source/dub/internal/sdlang/package.d source/dub/internal/sdlang/parser.d source/dub/internal/sdlang/symbol.d source/dub/internal/sdlang/token.d source/dub/internal/sdlang/util.d source/dub/internal/utils.d source/dub/internal/vibecompat/core/file.d source/dub/internal/vibecompat/core/log.d source/dub/internal/vibecompat/data/json.d source/dub/internal/vibecompat/data/serialization.d source/dub/internal/vibecompat/data/utils.d source/dub/internal/vibecompat/inet/path.d source/dub/internal/vibecompat/inet/url.d source/dub/package_.d source/dub/packagemanager.d source/dub/packagesupplier.d source/dub/packagesuppliers/fallback.d source/dub/packagesuppliers/filesystem.d source/dub/packagesuppliers/maven.d source/dub/packagesuppliers/package.d source/dub/packagesuppliers/packagesupplier.d source/dub/packagesuppliers/registry.d source/dub/platform.d source/dub/project.d source/dub/recipe/io.d source/dub/recipe/json.d source/dub/recipe/packagerecipe.d source/dub/recipe/sdl.d source/dub/semver.d source/dub/version_.d -vcolumns
(lldb) target create "../install/osx/bin/dmd"
Current executable set to '/Users/geod24/projects/dlang/install/osx/bin/dmd' (x86_64).
(lldb) settings set -- target.run-args  "-c" "-of.dub/build/application-debug-posix.osx.darwin-x86_64-dmd_2092-F034260D70B7CBC3C9D827805FB05AB9/dub.o" "-debug" "-g" "-w" "-version=DubUseCurl" "-version=DubApplication" "-version=Have_dub" "-Isource/" "source/app.d" "source/dub/commandline.d" "source/dub/compilers/buildsettings.d" "source/dub/compilers/compiler.d" "source/dub/compilers/dmd.d" "source/dub/compilers/gdc.d" "source/dub/compilers/ldc.d" "source/dub/compilers/utils.d" "source/dub/dependency.d" "source/dub/dependencyresolver.d" "source/dub/description.d" "source/dub/dub.d" "source/dub/generators/build.d" "source/dub/generators/cmake.d" "source/dub/generators/generator.d" "source/dub/generators/sublimetext.d" "source/dub/generators/targetdescription.d" "source/dub/generators/visuald.d" "source/dub/init.d" "source/dub/internal/git.d" "source/dub/internal/libInputVisitor.d" "source/dub/internal/sdlang/ast.d" "source/dub/internal/sdlang/exception.d" "source/dub/internal/sdlang/lexer.d" "source/dub/internal/sdlang/package.d" "source/dub/internal/sdlang/parser.d" "source/dub/internal/sdlang/symbol.d" "source/dub/internal/sdlang/token.d" "source/dub/internal/sdlang/util.d" "source/dub/internal/utils.d" "source/dub/internal/vibecompat/core/file.d" "source/dub/internal/vibecompat/core/log.d" "source/dub/internal/vibecompat/data/json.d" "source/dub/internal/vibecompat/data/serialization.d" "source/dub/internal/vibecompat/data/utils.d" "source/dub/internal/vibecompat/inet/path.d" "source/dub/internal/vibecompat/inet/url.d" "source/dub/package_.d" "source/dub/packagemanager.d" "source/dub/packagesupplier.d" "source/dub/packagesuppliers/fallback.d" "source/dub/packagesuppliers/filesystem.d" "source/dub/packagesuppliers/maven.d" "source/dub/packagesuppliers/package.d" "source/dub/packagesuppliers/packagesupplier.d" "source/dub/packagesuppliers/registry.d" "source/dub/platform.d" "source/dub/project.d" "source/dub/recipe/io.d" "source/dub/recipe/json.d" "source/dub/recipe/packagerecipe.d" "source/dub/recipe/sdl.d" "source/dub/semver.d" "source/dub/version_.d" "-vcolumns"
(lldb) run
Process 11023 launched: '/Users/geod24/projects/dlang/install/osx/bin/dmd' (x86_64)
Serializing composite type BuildRequirements which has no serializable fields
Serializing composite type BuildOptions which has no serializable fields
source/dub/packagesuppliers/maven.d(82,10): Deprecation: module std.xml is deprecated - Will be removed from Phobos in 2.101.0. If you still need it, go to https://github.com/DigitalMars/undeaD
Process 11023 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x0000000100216c10 dmd`OTeop1e(int) at oper.d:341
   338 	bool  OTpost(OPER op)      { return op == OPpostinc || op == OPpostdec; }
   339 	ubyte OTeop0e(OPER op)     { return optab1[op] & _OTeop0e; }
   340 	ubyte OTeop00(OPER op)     { return optab1[op] & _OTeop00; }
-> 341 	ubyte OTeop1e(OPER op)     { return optab1[op] & _OTeop1e; }
   342 	ubyte OTsideff(OPER op)    { return optab1[op] & _OTsideff; }
   343 	bool  OTconv(OPER op)      { return op >= CNVOPMIN && op <= CNVOPMAX; }
   344 	ubyte OTlogical(OPER op)   { return optab2[op] & _OTlogical; }
Target 0: (dmd) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x0000000100216c10 dmd`OTeop1e(int) at oper.d:341
    frame #1: 0x0000000100264200 dmd`optelem(elem*, unsigned int) at cgelem.d:5633
    frame #2: 0x000000010026433e dmd`optelem(elem*, unsigned int) at cgelem.d:5657
    frame #3: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #4: 0x00000001002639c8 dmd`optelem(elem*, unsigned int) at cgelem.d:5417
    frame #5: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #6: 0x0000000100263bb5 dmd`optelem(elem*, unsigned int) at cgelem.d:5454
    frame #7: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #8: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #9: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #10: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #11: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #12: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #13: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #14: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #15: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #16: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #17: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #18: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #19: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #20: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #21: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #22: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #23: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #24: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #25: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #26: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #27: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #28: 0x0000000100263b90 dmd`optelem(elem*, unsigned int) at cgelem.d:5452
    frame #29: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #30: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #31: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #32: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #33: 0x0000000100263bb5 dmd`optelem(elem*, unsigned int) at cgelem.d:5454
    frame #34: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #35: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #36: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #37: 0x00000001002637a5 dmd`optelem(elem*, unsigned int) at cgelem.d:5353
    frame #38: 0x0000000100263bb5 dmd`optelem(elem*, unsigned int) at cgelem.d:5454
    frame #39: 0x0000000100263536 dmd`optelem(elem*, unsigned int) at cgelem.d:5268
    frame #40: 0x0000000100263536 dmd`optelem(elem*, unsigned int) at cgelem.d:5268
    frame #41: 0x0000000100263536 dmd`optelem(elem*, unsigned int) at cgelem.d:5268
    frame #42: 0x0000000100264455 dmd`doptelem(elem*, unsigned int) at cgelem.d:5713
    frame #43: 0x0000000100273446 dmd`blockopt(int) at blockopt.d:681
    frame #44: 0x000000010024ff36 dmd`writefunc2(Symbol*) at out.d:1367
    frame #45: 0x000000010024f95b dmd`writefunc(Symbol*) at out.d:1102
    frame #46: 0x00000001001dee82 dmd`FuncDeclaration_toObjFile(FuncDeclaration*, bool) at glue.d:1237
    frame #47: 0x00000001001ed305 dmd`toObjFile::ToObjFile::visit(FuncDeclaration*) at toobj.d:310
    frame #48: 0x000000010010236e dmd`FuncDeclaration::accept(Visitor*) at func.d:2564
    frame #49: 0x00000001001ee576 dmd`_D3dmd5toobj9toObjFileRCQw7dsymbol7DsymbolbZ9ToObjFile5visitMRCQCj9dtemplate16TemplateInstanceZ__T9__lambda2TQDiZQpMFQDqZv at toobj.d:843
    frame #50: 0x0000000100087dfc dmd`_D3dmd7dsymbol14foreachDsymbolFPSQBf4root5array__T5ArrayTCQCeQCd7DsymbolZQxMDFQvZvZv at dsymbol.d:98
    frame #51: 0x00000001001ee54b dmd`toObjFile::ToObjFile::visit(TemplateInstance*) at toobj.d:843
    frame #52: 0x00000001000b4ed6 dmd`TemplateInstance::accept(Visitor*) at dtemplate.d:7441
    frame #53: 0x00000001001ed267 dmd`toObjFile(Dsymbol*, bool) at toobj.d:1027
    frame #54: 0x00000001001dd0e0 dmd`genObjFile(Module*, bool) at glue.d:389
    frame #55: 0x0000000100158c87 dmd`_D3dmd4mars7tryMainFmPPxaKSQz7globals5ParamZi at mars.d:713
    frame #56: 0x0000000100159b6c dmd`_Dmain at mars.d:952
    frame #57: 0x00000001003078cf dmd`_D2rt6dmain212_d_run_main2UAAamPUQgZiZ6runAllMFZ9__lambda1MFZv + 31
    frame #58: 0x0000000100307779 dmd`_D2rt6dmain212_d_run_main2UAAamPUQgZiZ7tryExecMFMDFZvZv + 33
    frame #59: 0x0000000100307853 dmd`_D2rt6dmain212_d_run_main2UAAamPUQgZiZ6runAllMFZv + 139
    frame #60: 0x0000000100307779 dmd`_D2rt6dmain212_d_run_main2UAAamPUQgZiZ7tryExecMFMDFZvZv + 33
    frame #61: 0x00000001003076dd dmd`_d_run_main2 + 621
    frame #62: 0x000000010030745a dmd`_d_run_main + 190
    frame #63: 0x00000001000018ec dmd`main at mars.d:911
    frame #64: 0x00007fff6a9c1cc9 libdyld.dylib`start + 1
    frame #65: 0x00007fff6a9c1cc9 libdyld.dylib`start + 1
```

**Comment by WalterBright** [comment]
thank you, @Geod24 . This might be enough for me to find the problem, I'll check it out.

**Comment by WalterBright** [comment]
What would help is if you could put:
```
elem_print(e);
```
as line 5356 so I can see the AST being processed.

**Comment by Geod24** [comment]
@WalterBright : The previous post has all necessary instructions to reproduce the bug. Just try them out on `dub`.

**Comment by MoonlightSentinel** [comment]
> Running `dustmite` right now to find a reduced test case (but that might take some time)

Sorry for the delay. Dustmite found a reduction that doesn't run indefinetly but takes way to much time to compile:

```
struct Param
{
    bool verbose;
    bool vcg_ast;
    bool showColumns;
    bool vtls;
    bool vtemplates;
    bool vgc;
    bool vfield;
    bool vcomplex;
    int useDeprecated;
    bool stackstomp;
    bool useUnitTests;
    bool useInline;
    bool useDIP25;
    bool noDIP25;
    bool useDIP1021;
    bool release;
    uint debuglevel;
    void* debugids;

    uint versionlevel;
    void* versionids;

    const(char)[] defaultlibname;
    const(char)[] debuglibname;
    const(char) mscrtlib;

    void* moduleDeps;
    int messageStyle = 1;
}

struct Global
{
    Param params;
}
```

```
dmd -c -version=MARS -fPIC -m64 -Jgenerated/linux/release/64 dmd/globals.d

real    0m0.031s
user    0m0.016s
sys     0m0.016s
```
```
generated/linux/release/64/dmd -c -version=MARS -fPIC -m64 -Jgenerated/linux/release/64 dmd/globals.d

real    0m9.054s
user    0m9.031s
sys     0m0.016s
```

**Comment by WalterBright** [comment]
Thank you, @MoonlightSentinel 
You da man!
I'll try this out later today.

**Comment by WalterBright** [comment]
@MoonlightSentinel test case works now. A new problem has appeared: the windows tests are timing out. Log files give no clue where.

**Comment by Geod24** [comment]
Do you have a Win64 to reproduce ? Also you can try to run with `-j1` to make sure the last test showing up is the one timing out.
Your best bet is to add a debug commit to set this to 1:
https://github.com/dlang/dmd/blob/ca636dcb8a56b53b8951ea8e5fca1bc581af639f/test/run.d#L91

**Comment by MoonlightSentinel** [comment]
Running the test suite with enabled assertions:
```
 ... runnable\builtin.d              -L/OPT:NOICF (-inline -release -g -O)
Test runnable\builtin.d failed.  The logged output:

generated\windows\release\64\dmd.exe -conf= -m64 -Irunnable  -L/OPT:NOICF  -odtest\test_results\runnable -oftest\test_results\runnable\builtin_0.exe  runnable\builtin.d
test\test_results\runnable\builtin_0.exe
0x1.f9f8d9aea10fep-2
0x1.f9f8d9aea10fep-2
0x1.bd21aaf88dcfap-1
0x1.bd21aaf88dcfap-1
0x1.22fd752af75cdp-1
0x1.22fd752af75cdp-1
Success

generated\windows\release\64\dmd.exe -conf= -m64 -Irunnable  -L/OPT:NOICF -inline -odtest\test_results\runnable -oftest\test_results\runnable\builtin_1.exe  runnable\builtin.d
Assertion failed: (count += 1) < 20, file src\dmd\backend\cgcod.d, line 2058

==============================
Test runnable\builtin.d failed: expected rc == 0, exited with rc == 9
```

The tests timeouts without assertions.

**Comment by WalterBright** [comment]
> The tests timeouts without assertions.

The compiler is designed to run with assertions on. There's a PR to fix that: https://github.com/dlang/dmd/pull/10679

**Comment by WalterBright** [comment]
@MoonlightSentinel thanks for finding the case, I'll check it out.

**Comment by WalterBright** [comment]
@Geod24 @MoonlightSentinel What I'm asking for is for the *test runner* to print:
```
dmd builtin.d
error: timed out after NNNN seconds
```
The purpose of a test suite is to not simply fail when it doesn't pass, but to aid in finding *where* it failed. I've posted many grumpy messages about this in recent weeks.

https://issues.dlang.org/show_bug.cgi?id=20926

**Comment by WalterBright** [comment]
@MoonlightSentinel I am having problems reproing it, as it works on my machine. builtin.d is a pretty small file. Is it possible you can reduce builtin.d down? The problem has something to do with the special register calling convention Win64 uses. Thanks!

**Comment by WalterBright** [comment]
meanwhile, I should figure out why my build is different

**Comment by Geod24** [comment]
@WalterBright : The timeout is not handled by the test runner, but by the CI provider. The auto-tester has a 1h timeout, Azure has 2h, etc... In order to do what you ask for, we would need to move the timeout feature to the test runner.

**Comment by WalterBright** [comment]
Moving it to the test runner sounds good to me.

This is not the first nor will be the last time this problem crops up and wastes several peoples' time. I've asked Brad to add "test suite" as a component in Bugzilla so I can start filing issues for all the problems I've been having with near-useless log files, heisenbugs, undocumented tests, scroll bars that are too small to be grabbed with the mouse, etc.

Because, after all, if it isn't in bugzilla it'll never get fixed :-)

**Comment by WalterBright** [comment]
@MoonlightSentinel never mind, I was able to duplicate it.

**Comment by WalterBright** [comment]
Finally! Working now.

File: test/compilable/test20906.d [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+ */
+
+int test12()
+{
+    int x = 0;
+    int a = x && 1 / x;
+    int b = !x || 1 / x;
+    int c = x ? 1 / x : 1;
+    int d = !x ? 1 : 1 / x;
+    return a | b | c;
```
Comment (v1ne) [inline test/compilable/test20906.d:13]: Why don't you also use `d`?
```suggestion
    return a | b | c | d;
```

Comment (WalterBright) [reply test/compilable/test20906.d:13]: It doesn't matter for the test.

**Comment by WalterBright** [comment]
@CyberShadow this was passing CyberShadow/DAutoTest yesterday, now it is failing. What can we do to make it reliable?

**Comment by CyberShadow** [comment]
> What can we do to make it reliable?

Generally:

- Pin all dependencies
- Don't force-merge broken PRs
- Ensure build scripts are deterministic and don't have any race conditions which would affect the outcome

It doesn't look like any of the above apply in this case:

```
-----------------------------------------------------------
[1msrc/dmd/templateparamsem.d(37): [1;33mWarning: [mDdoc: function declaration has no parameter 'tp'
[1msrc/dmd/templateparamsem.d(37): [1;33mWarning: [mDdoc: function declaration has no parameter 'sc'
[1msrc/dmd/templateparamsem.d(37): [1;33mWarning: [mDdoc: function declaration has no parameter 'parameters'
```

Why does that error appear now but (presumably) not previously? A regression introduced by this PR? A non-deterministic bug in DMD? In any case it doesn't look like a problem with the CI itself.

**Comment by CyberShadow** [comment]
> In any case it doesn't look like a problem with the CI itself.

It looks like a problem with the interaction of the CI and the dlang.org build scripts. The CI observed that it had built this commit before, but when trying to test it, the website then stupidly tries to build DMD from source, but the wrong DMD version is checked out at that point.

**Comment by CyberShadow** [comment]
I think it should be fixed in https://github.com/CyberShadow/ae/commit/3dd32f1bd79e4dfcb33ad20a5702898e54283a12 . Thanks for this bug report. I manually queued a retest for this PR.

**Comment by WalterBright** [comment]
> Why does that error appear now but (presumably) not previously?

I have no idea. The file also looks correct to me.

**Comment by WalterBright** [comment]
> I think it should be fixed

Ah, wonderful! Thanks for the quick action!

**Review by andralex — APPROVED** [review]

**Pull request merged**